.clang-format

```diff
@@ -79,31 +81,68 @@ MacroBlockBegin: ''
 MacroBlockEnd:   ''
 MaxEmptyLinesToKeep: 1
 NamespaceIndentation: None
+ObjCBinPackProtocolList: Never
+ObjCBlockIndentWidth: 2
+ObjCSpaceAfterProperty: false
+ObjCSpaceBeforeProtocolList: true
 PenaltyBreakAssignment: 2
-PenaltyBreakBeforeFirstCallParameter: 100
+PenaltyBreakBeforeFirstCallParameter: 1
 PenaltyBreakComment: 300
 PenaltyBreakFirstLessLess: 120
 PenaltyBreakString: 1000
-PenaltyExcessCharacter: 1
+PenaltyBreakTemplateDeclaration: 10
+PenaltyExcessCharacter: 1000000
 PenaltyReturnTypeOnItsOwnLine: 200
 PointerAlignment: Left
-BasedOnStyle:    google
+RawStringFormats:
+  - Language:        Cpp
+    Delimiters:
+      - cc
+      - CC
+      - cpp
+      - Cpp
+      - CPP
+      - 'c++'
+      - 'C++'
+    CanonicalDelimiter: ''
+    BasedOnStyle:    google
+  - Language:        TextProto
+    Delimiters:
+      - pb
+      - PB
+      - proto
+      - PROTO
+    EnclosingFunctions:
+      - EqualsProto
+      - EquivToProto
+      - PARSE_PARTIAL_TEXT_PROTO
+      - PARSE_TEST_PROTO
+      - PARSE_TEXT_PROTO
+      - ParseTextOrDie
+      - ParseTextProtoOrDie
+    CanonicalDelimiter: ''
+    BasedOnStyle:    google
 ReflowComments:  true
-SortIncludes:    true
+SortIncludes:    false
```

Comment:
I think we still need to sort the includes and get rid of the fragile code that depends on the order of the includes

Reply:
good to know, I'm just not sure what the course of action would be if we set `SortIncludes` back to `true`. Do we hold off any clang-format feature until the fragile code is resolved? Who would we reach out to to solve our fragile code issue? How do we prevent having more fragile code like this in the future?

Reply:
> good to know, I'm just not sure what the course of action would be if we set SortIncludes back to true. Who would we reach out to solve our fragile code issue?

Have you tried reformatting and building the entire repository? It seems the only expected problem is `velocypack-aliases` (the solution is to delete this file or write a [workaround](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#IncludeCategories) for this file in clang-format). And in general, I think such problems are solved quite realistically if you want, I can try on the weekend.

> Do we hold off any clang-format feature until the fragile code is resolved? 

I think we should merge PR like that, then do a big reformat and add validation to the CI and script for developers.
Then we should try to fix the problem with the order of includes and do a big reformat for them too.

> How do we prevent having more fragile code like this in the future?

Regarding the non-occurrence of such code in the future, this is usually done by reviewing and building the code on jenkins. And as with any other code.
Of course, in theory, there is a small chance that someone wrote code that depends on the order of inclusion, and then someone else ran into a problem, but it seems too small?

Reply:
My previous reformat & build is outdated (78bb583, f4c3b3d, 8410968, f0f6424), so I decided to do it again today locally. I was able to successfully build, as long as the following conditions are met after the reformat:
1. Revert the `#include` sorting in `arangod/Replication2/ReplicatedLog/ILogParticipant.cpp` (related to `velocypack-aliases`)
2. Revert the `#include` sorting in `arangod/V8Server/v8-user-functions.cpp` (related to `velocypack-aliases`)
3. Revert the `#include` sorting in `arangod/V8Server/v8-replicated-logs.cpp` (related to `velocypack-aliases`)
4. Add `#include <string>` to `lib/Utilities/IsArangoExecutable.h`

As for the solution to `velocypack-aliases`, I am not sure if removing the file would be easy, since it is currently being referenced in a significant amount of files (let me know what you think).

As for the next steps for clang-format, I am thinking of doing the following in order:
1. Figure out what to do with the `velocypack-aliases` issue
2. Set `SortIncludes` back to `true`
3. Merge the new `.clang-format` file into `devel` (this PR)
4. Open & merge new PR to perform the big reformat of entire codebase (will most likely be done on Dec 11)
5. Merge https://github.com/arangodb/arangodb/pull/15083 to introduce CI validation & pre-commit githook for developers. 

I want to figure out the `velocypack-aliases` issue as Step 1 since it prevents a successful build. 

Let me know what you think about all this,

CCing @joerg84 and @gnusi

Reply:
I'd have nothing against removing `velocypack-aliases`, but as you said, it'd be a lot of work.

But I think it should be usable with sorting. One problem I can think of is that it might still be used in some *header* files, in which it can break in other files including those headers. If that's the case, removing it from those headers should solve the problem. Or is the reason here a different one?

Reply:
`arangod/Replication2/ReplicatedLog/ILogParticipant.cpp`

Before sort:
```
#include "Replication2/LoggerContext.h"
#include "Replication2/ReplicatedLog/LogCore.h"
#include "Replication2/ReplicatedLog/LogStatus.h"
#include "Replication2/ReplicatedLog/ReplicatedLogMetrics.h"
#include "Metrics/Gauge.h"

#include <Basics/Exceptions.h>
#include <Basics/StaticStrings.h>
#include <Basics/application-exit.h>
#include <velocypack/velocypack-aliases.h>
```

After sort:
```
#include <Basics/Exceptions.h>
#include <Basics/StaticStrings.h>
#include <Basics/application-exit.h>
#include <velocypack/velocypack-aliases.h>

#include "Metrics/Gauge.h"
#include "Replication2/LoggerContext.h"
#include "Replication2/ReplicatedLog/LogCore.h"
#include "Replication2/ReplicatedLog/LogStatus.h"
#include "Replication2/ReplicatedLog/ReplicatedLogMetrics.h"
```

None of these header files are directly referencing `velocypack-aliases`, but `LogStatus.h` is referencing the following velocypack elements:
```
#include <velocypack/Builder.h>
#include <velocypack/Slice.h>
```
From what I've seen the only thing in common that `Build.h`, `Slice.h` and `velocypack-aliases` have is that they all include `velocypack/velocypack-common.h`. But I don't see anything alarming with that. 

@goedderz, do you notice anything unusual from this diff?

Reply:
@aMahanna I did mean you can use a workaround like this
```
IncludeCategories:
  - Regex:           '<velocypack/velocypack-aliases.h>'
    Priority:        4
  - Regex:           '^<ext/.*\.h>'
    Priority:        2
  - Regex:           '^<.*\.h>'
    Priority:        1
  - Regex:           '^<.*'
    Priority:        2
  - Regex:           '.*'
    Priority:        3
```

Reply:
I reformated & rebuilt ArangoDB with that new `.clang-format` style change, but encountered new errors, regarding `arangod/V8Server/v8-user-functions.cpp`. The re-ordering of the `#include` lines shifted `#include "v8-user-functions.h"` to be **before** others:
![image](https://user-images.githubusercontent.com/43019056/145274523-809c0e44-c064-4026-8a38-1ff9a8f5dbda.png)

This caused multiple errors of the `use of undeclared identifier` nature. 

The developer of this file must have intentionally placed the corresponding header file of `v8-user-functions.cpp` to be after all of these required elements to avoid any `undeclared identifier` errors.

Even if we fixed this one via another `IncludeCategories` addition, there is no guarantee that we won't see the issue again for another file. This could fill up our `IncludeCategories` with multiple regex matches to various header file names, which I am not a fan of having, as it would become difficult to maintain.

Reply:
Just use -priotlrity, like
1 => -5
2 => -4
3 => -3
4 => -2

Then main category, that have 0 priority by default will be last

Reply:
Btw I thought in your case these includes (V8/*) should be in header file, no?

Reply:
> Btw I thought in your case these includes (V8/*) should be in header file, no?
This definitely should be the case here, but I think this change is beyond the scope of what we want to have here

Reply:
> Just use negative priority, like
1 => -4
2 => -3
3 => -2
4 => -1

Sorry could you give an example of what that would look like in `.clang-format`? I’m guessing what you mean is to modify the priorities of the existing `IncludeCategories`, but not 100% certain

Reply:
```
IncludeCategories:
  - Regex:           '<velocypack/velocypack-aliases.h>'
    Priority:        -1
  - Regex:           '^<ext/.*\.h>'
    Priority:        -3
  - Regex:           '^<.*\.h>'
    Priority:        -4
  - Regex:           '^<.*'
    Priority:        -3
  - Regex:           '.*'
    Priority:        -2
```

Reply:
Changed SortIncludes to false